File: sdk/python/src/spendguard/ids.py
"""ID and idempotency-key helpers.

Two flavors of identity are minted here:

  - Time-ordered IDs (`new_uuid7`) for one-shot operations whose
    identity does not need to survive retries — e.g., Handshake's
    workload_instance_id, RunContext.run_id when the caller hasn't
    set one.

  - **Content-derived** IDs for everything inside a Pydantic-AI
    `Model.request()` call. Pydantic-AI's Agent run loop will call
    `request()` again on transient provider failure (or when the
    framework's retry policy fires). The wrapper has no signal
    distinguishing "fresh call" from "retry of the same step" other
    than the inputs themselves. So we derive `step_id` / `llm_call_id`
    / decision-trace-id / idempotency_key from a stable hash of the
    messages + model_settings + run_id. Same logical call → same hash
    → same idempotency_key → sidecar cache hit + ledger UNIQUE
    collapse onto the first decision.

References: Trace Schema §3.4 (idempotency_key fallback);
Sidecar Architecture §6 (idempotency cache).
"""

from __future__ import annotations

import hashlib
import os
import secrets
import time
import uuid
from collections.abc import Sequence
from typing import Any


def new_uuid7() -> uuid.UUID:
    """Mint a UUIDv7 (RFC 9562 §5.7).

    Layout (128 bits, big-endian):
      - 48 bits unix epoch ms
      - 4 bits version (0b0111)
      - 12 bits random
      - 2 bits variant (0b10)
      - 62 bits random
    """
    ts_ms = int(time.time() * 1000) & ((1 << 48) - 1)
    rand_a = secrets.randbits(12)
    rand_b = secrets.randbits(62)

    # Compose:
    #   bits 127..80  unix_ts_ms (48)
    #   bits 79..76   version 0x7 (4)
    #   bits 75..64   rand_a (12)
    #   bits 63..62   variant 0b10 (2)
    #   bits 61..0    rand_b (62)
    value = (
        (ts_ms << 80)
        | (0x7 << 76)
        | (rand_a << 64)
        | (0b10 << 62)
        | rand_b
    )
    return uuid.UUID(int=value)


def derive_idempotency_key(
    *,
    tenant_id: str,
    session_id: str,
    run_id: str,
    step_id: str,
    llm_call_id: str,
    trigger: str,
) -> str:
    """Deterministic idempotency key for a trigger boundary.

    Same (tenant, session, run, step, llm_call, trigger) → same key.
    A retry of the SAME logical step within Pydantic-AI's run loop
    (e.g., transient HTTP failure) MUST reuse this so the sidecar's
    cache short-circuits + the ledger's UNIQUE returns Replay.

    Returns a hex-encoded 32-char (128-bit) string — short enough to
    fit comfortably in log lines, wide enough for collision resistance.
    """
    canonical = "\x1f".join(
        [
            "v1",
            tenant_id,
            session_id,
            run_id,
            step_id,
            llm_call_id,
            trigger,
        ]
    )
    digest = hashlib.blake2b(canonical.encode("utf-8"), digest_size=16).hexdigest()
    return f"sg-{digest}"


CallSignatureFn = "Callable[[Sequence[Any], Any], str]"
"""Type alias for a custom call-signature function (see SpendGuardModel)."""


def default_call_signature(
    messages: Sequence[Any],
    model_settings: Any | None,
) -> str:
    """Stable hash over the *content* of a Pydantic-AI Model.request() call.

    The output is a 32-char hex digest. Two `Model.request()`
    invocations with identical messages and model_settings produce the
    same digest — which is what makes idempotency survive Pydantic-AI's
    framework-level retry, where the wrapper sees `request()` re-entered
    with bit-identical inputs.

    Serialization strategy:
      - Pydantic-AI message types are pydantic v2 models; we use
        `.model_dump_json(exclude_none=True)` for canonical form when
        available.
      - `model_settings` is typically a TypedDict; we sort-key
        json.dumps it.
      - Anything else falls back to `repr()` — brittle but deterministic
        within a single Python session.

    Callers that need stronger guarantees (cross-version stability,
    cross-runtime portability) should pass a custom `call_signature_fn`
    to `SpendGuardModel`.
    """
    import json

    h = hashlib.blake2b(digest_size=16)
    h.update(b"v1:call:")
    for i, msg in enumerate(messages):
        h.update(f"|msg{i}|".encode("utf-8"))
        if hasattr(msg, "model_dump_json"):
            try:
                h.update(msg.model_dump_json(exclude_none=True).encode("utf-8"))
                continue
            except Exception:  # noqa: BLE001 — fall through to repr
                pass
        h.update(repr(msg).encode("utf-8"))
    h.update(b"|settings|")
    if model_settings is None:
        h.update(b"none")
    elif hasattr(model_settings, "model_dump_json"):
        try:
            h.update(
                model_settings.model_dump_json(exclude_none=True).encode("utf-8")
            )
        except Exception:  # noqa: BLE001
            h.update(repr(model_settings).encode("utf-8"))
    elif isinstance(model_settings, dict):
        h.update(
            json.dumps(model_settings, sort_keys=True, default=str).encode("utf-8")
        )
    else:
        h.update(repr(model_settings).encode("utf-8"))
    return h.hexdigest()


def derive_uuid_from_signature(signature: str, *, scope: str) -> uuid.UUID:
    """Derive a stable UUID (v4-shaped) from a content signature + scope.

    `scope` ("decision_id", "llm_call_id", etc.) namespaces the UUID
    so different identifier slots never collide for the same call.
    """
    digest = hashlib.blake2b(
        f"{scope}|{signature}".encode("utf-8"), digest_size=16
    ).digest()
    buf = bytearray(digest)
    buf[6] = (buf[6] & 0x0F) | 0x40  # version 4
    buf[8] = (buf[8] & 0x3F) | 0x80  # variant 10
    return uuid.UUID(bytes=bytes(buf))


def workload_instance_id() -> str:
    """Sidecar workload identity hint.

    The adapter ASSERTS this in handshake; the sidecar verifies against
    SO_PEERCRED + signed manifest (per Sidecar §5). For POC we read it
    from the SPENDGUARD_WORKLOAD_INSTANCE_ID env var; production
    deployments inject this via the platform.
    """
    return os.environ.get("SPENDGUARD_WORKLOAD_INSTANCE_ID", "")

File: sdk/typescript/src/ids.ts
// SpendGuard SDK — ID helpers (SLICE 6 / COV_S05_06).
//
// Mirrors `sdk/python/src/spendguard/ids.py`. Two flavours of identity are
// minted here:
//
//   - Time-ordered IDs (`newUuid7`) for one-shot operations whose identity
//     does not need to survive retries (handshake `workload_instance_id`,
//     `RunContext.run_id` when the caller has not set one).
//   - **Content-derived** IDs for everything inside a single LLM call. A
//     retry of the SAME logical step within an agent framework's run loop
//     MUST reuse the same key so the sidecar's idempotency cache + the
//     ledger's `UNIQUE` constraint collapse onto the first decision.
//
// Spec refs:
//   - design.md §4.6 (LOCKED ID helper surface)
//   - implementation.md §6 (`src/ids.ts`)
//   - review-standards.md §1.5 (cross-language byte-equivalence)
//   - tests.md §5.3 (cross-language fixture matrix)
//
// Cross-language invariant: `deriveIdempotencyKey({...})` MUST produce the
// same string as Python `spendguard.ids.derive_idempotency_key(**kwargs)`
// for any byte-identical input. Same for `deriveUuidFromSignature`. Tested
// against shared fixtures in `tests/ids.test.ts`.

import { randomBytes } from "node:crypto";
import { blake2b } from "@noble/hashes/blake2";

/**
 * Mint a UUIDv7 per RFC 9562 §5.7.
 *
 * Layout (128 bits, big-endian):
 *   - 48 bits unix epoch ms
 *   - 4 bits version (0b0111)
 *   - 12 bits random
 *   - 2 bits variant (0b10)
 *   - 62 bits random
 *
 * Two calls within the same ms are time-ordered to ms precision but
 * randomised within the ms slot. Returned in canonical 36-char hex form
 * (`xxxxxxxx-xxxx-7xxx-yxxx-xxxxxxxxxxxx`).
 */
export function newUuid7(): string {
  const tsMs = BigInt(Date.now()) & ((1n << 48n) - 1n);
  // 12 bits of rand_a
  const randA = randomBytes(2).readUInt16BE(0) & 0x0fff;
  // 62 bits of rand_b + 2-bit variant overwrite
  const randB = randomBytes(8);
  // Overwrite top 2 bits of byte 8 with variant 0b10
  randB[0] = (randB[0]! & 0x3f) | 0x80;

  // bits 127..80  ts_ms (48)
  // bits 79..76   version 0x7
  // bits 75..64   rand_a (12)
  // bits 63..62   variant 0b10 (already set in randB)
  // bits 61..0    rand_b
  const hi = (tsMs << 16n) | BigInt(randA);
  const buf = Buffer.alloc(16);
  buf.writeBigUInt64BE(hi, 0);
  // Overwrite top nibble of byte 6 with version 7
  buf[6] = (buf[6]! & 0x0f) | 0x70;
  randB.copy(buf, 8);

  return [
    buf.toString("hex", 0, 4),
    buf.toString("hex", 4, 6),
    buf.toString("hex", 6, 8),
    buf.toString("hex", 8, 10),
    buf.toString("hex", 10, 16),
  ].join("-");
}

/**
 * Fixed canonical-form separator.
 *
 * Mirrors Python `derive_idempotency_key`: ASCII Unit Separator (\x1f) joins
 * the fields. The leading `"v1"` schema tag is included.
 */
const FIELD_SEP = "\x1f";

/**
 * Deterministic idempotency key for a trigger boundary.
 *
 * Same `(tenantId, sessionId, runId, stepId, llmCallId, trigger)` →
 * same key. A retry of the SAME logical step within an agent framework's
 * run loop MUST reuse this so the sidecar's cache short-circuits + the
 * ledger's `UNIQUE` returns Replay.
 *
 * Returns `"sg-"` + 32 hex chars (128-bit BLAKE2b digest).
 *
 * **Cross-language gate (P0 — review-standards §2.2)**: the TS output for
 * any given input is byte-identical to Python
 * `derive_idempotency_key(**kwargs)` (which uses `hashlib.blake2b(...,
 * digest_size=16)`). The implementation here calls the audited
 * `@noble/hashes` BLAKE2b primitive with `dkLen: 16` — the same 128-bit
 * personalised-output mode Python's hashlib exposes. Verified byte-for-byte
 * against `derive_idempotency_key(**kwargs)` for 7 fixtures in
 * `tests/ids.test.ts` (FX1–FX7).
 */
export function deriveIdempotencyKey(args: {
  tenantId: string;
  sessionId: string;
  runId: string;
  stepId: string;
  llmCallId: string;
  trigger: string;
}): string {
  const canonical = [
    "v1",
    args.tenantId,
    args.sessionId,
    args.runId,
    args.stepId,
    args.llmCallId,
    args.trigger,
  ].join(FIELD_SEP);
  // BLAKE2b-128 (16-byte digest) — byte-equivalent to Python
  // `hashlib.blake2b(..., digest_size=16).hexdigest()`. Locked by
  // review-standards §2.2 cross-language P0 gate.
  const digest = blake2b(new TextEncoder().encode(canonical), { dkLen: 16 });
  let hex = "";
  for (let i = 0; i < digest.length; i++) {
    hex += digest[i]!.toString(16).padStart(2, "0");
  }
  return `sg-${hex}`;
}

/**
 * Derive a stable UUID (v4-shaped) from a content signature + scope namespace.
 *
 * Used for content-derived `decision_id` / `llm_call_id` slots where a retry
 * of the SAME `Model.request()` MUST produce the same UUID across retries.
 * Scope ("decision_id", "llm_call_id", etc.) namespaces the output so
 * different identifier slots never collide for the same signature.
 *
 * Same `(signature, scope)` → same UUID. Byte-equivalent to Python
 * `spendguard.ids.derive_uuid_from_signature(signature, scope=scope)`:
 * BLAKE2b-128 over `f"{scope}|{signature}"`, with RFC 4122 v4 version (0x4)
 * and variant (0b10) nibbles overlaid on bytes 6 + 8. Verified against
 * 5 fixtures in `tests/ids.test.ts` (FXU1–FXU5).
 */
export function deriveUuidFromSignature(signature: string, args: { scope: string }): string {
  const canonical = `${args.scope}|${signature}`;
  const digest = blake2b(new TextEncoder().encode(canonical), { dkLen: 16 });
  const buf = Buffer.from(digest);
  // Version 4 + variant 10
  buf[6] = (buf[6]! & 0x0f) | 0x40;
  buf[8] = (buf[8]! & 0x3f) | 0x80;
  return [
    buf.toString("hex", 0, 4),
    buf.toString("hex", 4, 6),
    buf.toString("hex", 6, 8),
    buf.toString("hex", 8, 10),
    buf.toString("hex", 10, 16),
  ].join("-");
}

/**
 * Sidecar workload identity hint.
 *
 * The adapter asserts this in handshake; the sidecar verifies against
 * SO_PEERCRED + signed manifest. Reads `SPENDGUARD_WORKLOAD_INSTANCE_ID`
 * env var; returns empty string when unset.
 */
export function workloadInstanceId(): string {
  return process.env.SPENDGUARD_WORKLOAD_INSTANCE_ID ?? "";
}
